Add unspecified-encoding checker #3826

DanielNoord opened the pull request:
- [x] Add yourself to CONTRIBUTORS if you are a new contributor.
- [x] Add a ChangeLog entry describing what your PR does.
- [x] If it's a new feature, or an important bug fix, add a What's New entry in
      `doc/whatsnew/<current release.rst>`.
- [x] Write a good description on what the PR does.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |


## Description

This adds an unspecified-encoding checker that adds a warning whenever open() is called without an explicit encoding argument.
This closes #3826.

I stumbled upon this issue today while I experienced the importance of adding `encoding` to `open()`. This is my first contribution to pylint so please let me know if anything needs to be changed or improved!

Pierre-Sassoulas commented on ChangeLog:
```diff
@@ -6,8 +6,9 @@ What's New in Pylint 2.10.0?
 
-..
-  Put new features here and also in 'doc/whatsnew/2.10.rst'
+* Added ``unspecified-encoding``: Emitted when open() is called without specifying an encoding
```
```suggestion
..
  Put new features here and also in 'doc/whatsnew/2.10.rst'
  
* Added ``unspecified-encoding``: Emitted when open() is called without specifying an encoding
```

Pierre-Sassoulas commented on pylint/checkers/stdlib.py:
```diff
@@ -425,6 +426,13 @@ class StdlibChecker(DeprecatedMixin, BaseChecker):
+            "Using open without explicitly specifying an encoding",
+            "unspecified-encoding",
+            "It is better to specify an encoding when opening documents. "
+            "Failing to do so can create problems when opening documents on other systems. "
```
```suggestion
            "Using the system default implicitly can create problems on other operating systems. "
```

Pierre-Sassoulas commented on pylint/checkers/stdlib.py:
```diff
@@ -573,6 +584,17 @@ def _check_open_mode(self, node):
             ):
                 self.add_message("bad-open-mode", node=node, args=mode_arg.value)
 
+    def _check_open_encoded(self, node):
```
```suggestion
    def _check_open_encoded(self, node: astroid.Call) -> None:
```

Pierre-Sassoulas reviewed (changes requested):
Thank you for handling this issue. This looks great. I made some small comments but this is definitely going in ``2.10.0`` :).

Pierre-Sassoulas commented on tests/functional/u/unspecified-encoding_py3.py:
```diff
@@ -0,0 +1,7 @@
+NAME = "foo.bar"
+open(NAME, "wb", encoding="utf-8")
+open(NAME, "w", encoding=None)  # [unspecified-encoding]
+open(NAME, "rb")  # [unspecified-encoding]
```
Let's add the example given in the issue :)
```suggestion
open(NAME, "rb")  # [unspecified-encoding]

with open(filename) as f:  # [unspecified-encoding]
    ...

with open(filename, encoding=None) as f:  # [unspecified-encoding]
    ...


with open(filename, encoding="utf8", errors="surrogateescape") as f:
    ...

locale_encoding = getattr(io, "LOCALE_ENCODING", None)
with open(filename, encoding=locale_encoding) as f:
    ...
```
Also we should probably check more use case to handle other open method (method named open, other method from io, etc.)

DanielNoord commented:
@Pierre-Sassoulas I merged all your suggestions + added checks for `b` mode, which does not need an encoding.

I do not really have any experience with other `open` methods. Perhaps you can add some additional test cases your self, or give me a link to some documentation on `open` methods you would like to see added.

Pierre-Sassoulas commented:
I think we should handle [``io.open``](https://docs.python.org/3/library/io.html#io.open), it's just an alias, but also [``io.open_code`` which is implicitly "rb"](https://docs.python.org/3/library/io.html#io.open_code) and make sure that we don't have false positive for it.

DanielNoord commented:
Added the check for `io.open` and `io.open_code`. 

However, the test suite is failing on some files which have not been edited by these commits while it is up to date with the main branch (besides the latest commit to issue forms).
Not sure what is going wrong here. Should I fix that in this PR? I don't immediately know where these errors are coming from.

Pierre-Sassoulas commented:
> Not sure what is going wrong here. 

Well what happened is you discovered style problem inside pylint with your new checker 😄 The new problem in pylint should be fixed in this merge request, we can discuss them here. Maybe it'll fix strange bug that happened only on some particular OS 🎉 

Or it can be problem in existing functional tests that are now detected. You can disable those new messages in the existing tests so there is no change in the expected output, or update the output if you think this make more sense.

Pierre-Sassoulas reviewed (approved):
👍

Pierre-Sassoulas commented:
Always happy to help :) Also don't stress out, this won't be merged before 2.10, so not before 2.9.6 is out 🙂 

Regarding the latest change to fix the tests, it looks most of the changes are in functional tests. Please don't modify the functional test too much, you can disable the new checker with ``pylint: disable=unspecified-encoding`` at the top of the file (avoid creating a newline as the expected output will change completely if you add a line). The formatting changes seems like they are automated by your IDE and need to be reverted as having strange formattign is also part of what we're testing (``j+1;lambda ...`` for example). Sorry for not being clearer in my earlier comment.

DanielNoord commented:
Ah shit, I just changed some of the tests to include `encodings` and fixed line-numbering in the test data. Do you want me to remove those and add the override? Or is adding `encodings` to fix the errors also fine?
For me (Mac) the tests are also throwing a number of `c-extension-no-member`, `no-member` and `no-name-in-module` errors. Are those also related to the changes introduced by this PR? I don't think so, right?

Pierre-Sassoulas commented:
A mix of disabling, changing the code and checking that the warning is raised is fine, it should not be one sided though, more like a third of each. (To be faire disable or checking that it's raised is an easier change but you already did the work of changing the code so... 😄)

I don't think the test suite run in the CI for mac right now as we have linux and windows only.  So, its possible that we have some existing problem on mac.  We supposed mac/linux were close enough and not worth the added pipeline time. Apparently it's not. You can check on the master branch for the test that are already failing while we fix this.

DanielNoord commented:
Hmm, after fixing my formatter the tests are now also raising a number of issues with the code of pylint itself.
I guess `open()` is used often, who would have thought... :)

I would be very glad to fully fix all errors, but this won't be done tonight. Seeing as pylint has such a good test coverage already, I would feel bad if my contribution would lead to an increase in skipped checks. I will try and group some of the changes in similar commits for easier review. We can always squash those before the final merge.

cdce8p commented on pylint/checkers/stdlib.py:
```diff
@@ -44,7 +45,7 @@
 from pylint.interfaces import IAstroidChecker
 
-OPEN_FILES = {"open", "file"}
+OPEN_FILES_MODE = {"open", "file"}
```
```suggestion
OPEN_FILES_MODE = ("open", "file")
```
Why not use a tuple here? A set is a bit over the top for just two entries.

DanielNoord replied:
I mainly did this because is was already as set. For example, on line 53 `ENV_GETTERS` is even a set with only one entry.
Want me to change both to a tuple? Or change this to a tuple and line 53 to a string?

cdce8p replied:
Let's stick with a tuple for `ENV_GETTERS`. That way the `if` statement doesn't need to change if we ever need to add additional entries.

cdce8p commented on pylint/checkers/stdlib.py:
```diff
@@ -494,8 +503,12 @@ def visit_call(self, node):
-                    if getattr(node.func, "name", None) in OPEN_FILES:
+                    if getattr(node.func, "name", None) in OPEN_FILES_MODE:
                         self._check_open_mode(node)
+                    if getattr(node.func, "name", None) in {"open"} or getattr(
+                        node.func, "attrname", None
+                    ) in {"open"}:
```
```suggestion
                    if getattr(node.func, "name", None) in ("open",) or getattr(
                        node.func, "attrname", None
                    ) in ("open",):
```
Same here. Tuples will work just fine.

DanielNoord replied:
Wouldn't a comparison with the string be fine as well? Or is comparison to a tuple better performance wise?

cdce8p replied:
Same reason as above. Although a string comparison might be possible, it would require us to change the `if` statement if we ever add additional entries. It might even be a good idea to create a new constant for `("open",)`.

cdce8p commented on pylint/checkers/stdlib.py:
```diff
@@ -494,8 +503,12 @@ def visit_call(self, node):
-                    if getattr(node.func, "name", None) in OPEN_FILES:
+                    if getattr(node.func, "name", None) in OPEN_FILES_MODE:
                         self._check_open_mode(node)
+                    if getattr(node.func, "name", None) in {"open"} or getattr(
+                        node.func, "attrname", None
+                    ) in {"open"}:
```
```suggestion
                    if (
                        isinstance(node.func, astroid.Name) and node.func.name in ("open",)
                        or isinstance(node.func, astroid.Attribute) and node.func.attrname in ("open",)
                    ):
```
Furthermore, it might be worth considering to replace the `getattr` calls with `isinstance` once.

cdce8p commented on doc/whatsnew/2.10.rst:
```diff
@@ -12,7 +12,9 @@ Summary -- Release highlights
 
+* Added ``unspecified-encoding``: Emitted when open() is called without specifying an encoding
 
+  Closes #3826
```
```suggestion
  Closes #3826

```
Just a small suggestion: Can you add an empty line her? Usually there should be two empty ones before a new section.

cdce8p reviewed (commented):
What a first PR 🔥 I like it a lot! Did a test run and it already worked beautifully.
As this will go in `2.10`, there is still some time. So you don't need to hurry.

DanielNoord commented:
Almost made it work. Again, sorry for taking so much of your time...

Action at [personal repo](https://github.com/DanielNoord/pylint/pull/1) shows that the PR succeeds for CPython 3.8 =<, but fails for 3.7, 3.6 and pypy. I thought appending `_py38` would make it automatically skip for 3.6 and 3.7 but I guess it is not that simple.
Furthermore, it is only one line that is failing (line 34) as `io.open_code()` was only added in 3.8. Perhaps there is a way to only make it skip that line for 3.6 and 3.7?

cdce8p commented:
You would need to create a file like this one: [generic_alias_typing.rc](https://github.com/PyCQA/pylint/blob/c04f92ef68e5ea779a60bfddb91dc677c5470fd0/tests/functional/g/generic_alias/generic_alias_typing.rc). With the same name as the test file, only `rc` as extension.

One more thing, I would appreciate if you wouldn't force push all your new changes. That makes it really difficult to review them. We can squash merge the PR in the end so it doesn't make a difference if there are 10+ commits. Usually a force push should only be used when doing a rebase to catch up with the latest changes on `main`.

DanielNoord commented:
I'm really sorry! I thought you might want a commit history that represents a possible final history.
Will refrain from force pushing in the future!

cdce8p commented on pylint/checkers/stdlib.py:
```diff
@@ -494,8 +504,19 @@ def visit_call(self, node):
+                    if (
+                        isinstance(node.func, astroid.Name)
+                        and node.func.name in OPEN_FILES_ENCODING
+                    ) or (
+                        isinstance(node.func, astroid.Attribute)
```
```suggestion
                        or isinstance(node.func, astroid.Attribute)
```
The parentheses aren't really necessary as `and` binds tighter than `or`.

cdce8p commented on pylint/lint/pylinter.py:
```diff
@@ -617,8 +619,9 @@ def set_option(self, optname, value, action=None, optdict=None):
                 except KeyError:
                     meth = self._bw_options_methods[optname]
                     warnings.warn(
-                        "%s is deprecated, replace it by %s"
-                        % (optname, optname.split("-")[0]),
+                        "{} is deprecated, replace it by {}".format(
+                            optname, optname.split("-")[0]
+                        ),
```
As this isn't related, it would be best to exclude the change here. Even if technically it's an improvement.

DanielNoord replied:
Want me to create another commit/PR to do this? Or just let it be?

cdce8p replied:
If you really like to, you can create another one. But if you ask me, I would let it be. It doesn't harm anybody and is still valid code.

cdce8p reviewed (approved):
Looks good 🚀
Left some last comments

Pierre-Sassoulas commented:
Haaa. I made a mistake I should have waited for 2.9.6 to be released before merging.

cdce8p commented:
> Haaa. I made a mistake I should have waited for 2.9.6 to be released before merging.

In that case, what do you think about releasing `2.10` next? As far as I've seen, there aren't any major bugfixes that need to go into `2.9.6`. All can move to `2.10`

DanielNoord commented:
Also note that I didn't have time to merge @cdce8p's last comment

Pierre-Sassoulas commented:
> Also note that I didn't have time to merge @cdce8p's last comment

Don't worry, I applied the suggestion :)

I'm going to release 2.9.6 with #4765  and we'll start to merge the 2.10 features for the next release after that. I moved the remaining MR in 2.9.6 to 2.10.1.
